From 5bb6dc8449d37c3d57c774e0669b66794104e44d Mon Sep 17 00:00:00 2001
From: Qiang Yu <yuq825@gmail.com>
Date: Tue, 11 Feb 2020 10:32:38 +0800
Subject: [PATCH 073/102] lima: always add texture bo to submit

No matter texture desc change, we need to add texture to submit.
Otherwise texture may be freed before submit finish.

Signed-off-by: Qiang Yu <yuq825@gmail.com>
---
 src/gallium/drivers/lima/lima_draw.c    |  3 +--
 src/gallium/drivers/lima/lima_texture.c | 13 +++++++++++--
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/src/gallium/drivers/lima/lima_draw.c b/src/gallium/drivers/lima/lima_draw.c
index b0eb479986b..933ca929181 100644
--- a/src/gallium/drivers/lima/lima_draw.c
+++ b/src/gallium/drivers/lima/lima_draw.c
@@ -1518,8 +1518,7 @@ lima_draw_vbo_update(struct pipe_context *pctx,
       ctx->const_buffer[PIPE_SHADER_FRAGMENT].dirty = false;
    }
 
-   if (ctx->dirty & LIMA_CONTEXT_DIRTY_TEXTURES)
-      lima_update_textures(ctx);
+   lima_update_textures(ctx);
 
    lima_pack_render_state(ctx, info);
    lima_pack_plbu_cmd(ctx, info);
diff --git a/src/gallium/drivers/lima/lima_texture.c b/src/gallium/drivers/lima/lima_texture.c
index 7f2f81266b6..793b8f31b8f 100644
--- a/src/gallium/drivers/lima/lima_texture.c
+++ b/src/gallium/drivers/lima/lima_texture.c
@@ -95,8 +95,6 @@ lima_texture_desc_set_res(struct lima_context *ctx, lima_tex_desc *desc,
       layout = 0;
    }
 
-   lima_submit_add_bo(ctx->pp_submit, lima_res->bo, LIMA_SUBMIT_BO_READ);
-
    uint32_t base_va = lima_res->bo->va;
 
    /* attach first level */
@@ -263,6 +261,17 @@ lima_update_textures(struct lima_context *ctx)
    if (!lima_tex->num_samplers || !lima_tex->num_textures)
       return;
 
+   /* we always need to add texture bo to submit */
+   for (int i = 0; i < lima_tex->num_samplers; i++) {
+      struct lima_sampler_view *texture = lima_sampler_view(lima_tex->textures[i]);
+      struct lima_resource *rsc = lima_resource(texture->base.texture);
+      lima_submit_add_bo(ctx->pp_submit, rsc->bo, LIMA_SUBMIT_BO_READ);
+   }
+
+   /* do not regenerate texture desc if no change */
+   if (!(ctx->dirty & LIMA_CONTEXT_DIRTY_TEXTURES))
+      return;
+
    unsigned size = lima_tex_list_size;
    for (int i = 0; i < lima_tex->num_samplers; i++) {
       struct lima_sampler_view *texture = lima_sampler_view(lima_tex->textures[i]);
-- 
2.24.1

